File: python/Day10/day10.py
def twist(nums, start, move):
    nums = nums[::]
    end = start + move
    if end > len(nums):
        new = nums[start:] + nums[0:end % len(nums)]
    else:
        new = nums[start:end]
    new = new[::-1]
    for i in range(len(new)):
        nums[(start + i) % len(nums)] = new[i]
    return nums


inpt = list(map(int, "63,144,180,149,1,255,167,84,125,65,188,0,2,254,229,24".split(',')))
start, skip = 0, 0
data = list(range(256))

for move in inpt:
    data = twist(data, start, move)
    start = (start + move + skip) % len(data)
    skip += 1

print('Part 1:', data[0]*data[1])

inpt = list(map(ord, "63,144,180,149,1,255,167,84,125,65,188,0,2,254,229,24")) + [17, 31, 73, 47, 23]
data = list(range(256))
start, skip = 0, 0
for _ in range(64):
    for move in inpt:
        data = twist(data, start, move)
        start = (start + move + skip) % len(data)
        skip += 1

hashes = []
for i in range(0,16):
    num = 0
    for n in data[16 * i: 16 * (i + 1)]:
        num ^= n
    hashes.append(num)

print('Part 2:', ''.join(list(map(lambda x: "{:02x}".format(x), hashes))))

File: kotlin/src/main/AoC2017/day10.kt
package main.AoC2017

import java.io.File

fun main(args: Array<String>){

    val input = File("inputs/2017/day10.in").readText()
    println("Part 1: ${solvePart1(input)}")
    println("Part 2: ${getKnotHash(input)}")

}

fun solvePart1(input: String) : Int{
    val moves = input.split(",").map { it.toInt() }
    var data = IntRange(0, 255).toList()
    var start = 0
    var skip = 0

    for(move in moves){
        data = twist(data, start, move)
        start = (start + move + skip++) % data.size
    }
    return data[0] * data[1]
}

fun getKnotHash(input: String): String{
    val moves = input.map{it.toInt()}.toList() + listOf(17, 31, 73, 47, 23)
    var data = IntRange(0, 255).toList()
    var start = 0
    var skip = 0

    for(i in 0 until 64) {
        for (move in moves) {
            data = twist(data, start, move)
            start = (start + move + skip++) % data.size
        }
    }
    
    return data.chunked(16)
            .map{it.reduce{x, y -> x xor y}}
            .map{"%02x".format(it)}
            .joinToString("")
}


fun twist(data: List<Int>, start: Int, offset: Int ) : List<Int>{
    val end = start + offset
    val toReverse = if(end > data.size){
        (data.subList(start, data.size) + data.subList(0, end % data.size)).toMutableList()
    }else {
        data.subList(start, end).toMutableList()
    }
    val newList = data.toMutableList()
    for(i in toReverse.indices){
        newList[(start + i) % newList.size] = toReverse[toReverse.size - 1 - i]
    }
    return newList.toList()
}


